mesh/reward_test.go

```diff
@@ -211,57 +211,85 @@ func TestMesh_updateStateWithLayer(t *testing.T) {
 	mesh, atxDB := getMeshWithMapState(t, "t1", s)
 	defer mesh.Close()
 
-	for i := 1; i <= numOfLayers; i++ {
-		createLayer(t, mesh, types.NewLayerID(uint32(i)), numOfBlocks, maxTxs, atxDB)
-		l, err := mesh.GetLayer(types.NewLayerID(uint32(i)))
-		assert.NoError(t, err)
-		mesh.ValidateLayer(l)
+	startLayer := types.GetEffectiveGenesis()
+
+	for i := 0; i < numOfLayers; i++ {
+		layerID := startLayer.Add(uint32(i))
+		createLayer(t, mesh, layerID, numOfBlocks, maxTxs, atxDB)
+		l, err := mesh.GetLayer(layerID)
+		require.NoError(t, err)
+		mesh.ValidateLayer(context.TODO(), l)
 	}
 
 	s2 := &MockMapState{Rewards: make(map[types.Address]*big.Int)}
 	mesh2, atxDB2 := getMeshWithMapState(t, "t2", s2)
 
-	// this should be played until numOfLayers -1 if we want to compare states
-	for i := 1; i <= numOfLayers-1; i++ {
-		blockIds := copyLayer(t, mesh, mesh2, atxDB2, types.NewLayerID(uint32(i)))
-		mesh2.HandleValidatedLayer(context.TODO(), types.NewLayerID(uint32(i)), blockIds)
+	// this should be played until numOfLayers-1 if we want to compare states
+	for i := 0; i < numOfLayers-1; i++ {
+		layerID := startLayer.Add(uint32(i))
+		blockIds := copyLayer(t, mesh, mesh2, atxDB2, layerID)
+		mesh2.HandleValidatedLayer(context.TODO(), layerID, blockIds)
 	}
+
 	// test states are the same when one input is from tortoise and the other from hare
-	assert.Equal(t, s.Txs, s2.Txs)
+	require.NotEqual(t, s.Txs, s2.Txs)
+
+	layerIDFinal := startLayer.Add(uint32(numOfLayers - 1))
+	copyLayer(t, mesh, mesh2, atxDB2, layerIDFinal)
+	l, err := mesh.GetLayer(layerIDFinal)
+	require.NoError(t, err)
+	mesh2.ValidateLayer(context.TODO(), l)
+
+	// test states are the same when one input is from tortoise and the other from hare
+	require.Equal(t, s.Txs, s2.Txs)
```

Comment:
why not do `for i := 0; i < numOfLayers; i++` in line 228 and `require.Equal(t, s.Txs, s2.Txs)` directly?
why care about this one last layer difference?

Reply:
the test is more thorough if it isolates the final layer application: the states are _not_ equal before the final layer is applied, and they _are_ equal after.